Core/include/Acts/Geometry/Portal.hpp

```diff
@@ -0,0 +1,234 @@
+                      const Logger& logger = getDummyLogger());
+
+  /// Resolve the volume for a 3D position and a direction
+  /// The @p direction is used to select the right portal link, if it is set.
+  /// In case no link is found in the specified direction, a `nullptr` is
+  /// returned.
+  /// @param gctx The geometry context
+  /// @param position The 3D position
+  /// @param direction The direction
+  /// @return The target volume (can be `nullptr`)
+  Result<const TrackingVolume*> resolveVolume(const GeometryContext& gctx,
+                                              const Vector3& position,
+                                              const Vector3& direction) const;
+
+  /// Set a link on the portal into the slot specified by the direction.
+  /// @note The surface associated with @p link must be logically equivalent
+  ///       to the one of the link that's already set on the portal.
+  /// @param gctx The geometry context
+  /// @param direction The direction
+  /// @param link The link to set
+  void setLink(const GeometryContext& gctx, Direction direction,
+               std::unique_ptr<PortalLinkBase> link);
+
+  /// Helper function create a trivial portal link based on a surface.
+  /// @param gctx The geometry context
+  /// @param direction The direction of the link to create
+  /// @param surface The surface
+  /// @note The @p surface must be logically equivalent
+  ///       to the one of the link that's already set on the portal.
+  /// @param volume The target volume
+  void setLink(const GeometryContext& gctx, Direction direction,
+               std::shared_ptr<RegularSurface> surface, TrackingVolume& volume);
+
+  /// Get the link associated with the @p direction. Can be null if the associated link is unset.
+  /// @param direction The direction
+  /// @return The link (can be null)
+  const PortalLinkBase* getLink(Direction direction) const;
+
+  /// Returns true if the portal is valid, that means it has at least on
+  /// non-null link associated.Portals can be in an invalid state after they get
+  /// merge or fused with other portals.
+  /// @return True if the portal is valid
+  bool isValid() const;
+
+  /// Access the portal surface that is shared between the two links
+  /// @return The portal surface
+  const RegularSurface& surface() const;
+
+ private:
+  /// Helper to check surface equivalence without checking material status. This
+  /// is needed because we allow fusing portals with surfaces that are
+  /// equivalent but one of them has material while the other does not. The
+  /// normal surface comparison would determine these surfaces as not
+  /// equivalent.
+  /// @param gctx The geometry context
+  /// @param a The first surface
+  /// @param b The second surface
+  /// @return True if the surfaces are equivalent
+  static bool isSameSurface(const GeometryContext& gctx, const Surface& a,
+                            const Surface& b);
```

Comment:
I wonder if this should live in the `Surface` class / header. Potentially such an equality check could be useful to other components

Reply:
We do have an equality operator, but that also checks if both surfaces have material. In this case, I explicitly want to allow the case where only one surface has material. In fact, the class will refuse fusing portals that both have material already.